q2_cutadapt/_demux.py

```diff
@@ -119,6 +119,15 @@ def _write_empty_fastq_to_mux_barcode_in_seq_fmt(seqs_dir_fmt):
         seqs_dir_fmt.file.write_data(fastq, FastqGzFormat)
 
 
+def _clean_empty(seqs):
+    """Clean up empty files output by Cutadapt"""
+    for _, fastqgz in seqs.sequences.iter_views(FastqGzFormat):
+        with gzip.open(str(fastqgz), 'rb') as f:
+            data = f.read(1)
+        if len(data) == 0:
+            (seqs.path / fastqgz.path).unlink()
```

Comment:
@ebolyen - can you lay some eyes on this when you get a chance? Looking for some feedback or pitfalls associated with checking the contents of a gzip file this way.

Reply:
There shouldn't be any issue with this, I might recommend `f.peek(1)` or `f.read1(1)`, which should guarantee only a single read call is made which is sufficient to check for empty, but really that doesn't matter at all.

Reply:
Awesome, thanks for looking, @ebolyen!

Reply:
@andrewsanchez, I just realized that this approach might produce invalid results in certain PE cases. If all of the reads are filtered out in a particular direction but not the other, we can wind up with missing directions in a pair. For example:

## fwd muxed
```
@id1
AAAA
+
zzzz
```

## rev muxed
```
@id1
ACGT
+
zzzz
```

## barcodes
```
id    barcode
s1    AAAA
```

This would produce an empty s1 R1 file (because the barcode gets trimmed out by `cutadapt`), but the s1 R2 file wouldn't be empty. When `_clean_empty` hits the R1 it'll remove the empty file, but leave behind the R2. This is a problem. We have the ability to `view` a demux output's `.manifest`, but I don't think you'll have one at this stage. I need to think on this a bit...

Reply:
Well hot dog!

https://github.com/qiime2/q2-types/blob/a547c8b075b66aca80f7f69a3bc7f95368aec32a/q2_types/per_sample_sequences/_format.py#L304-L329

Apparently we are in good shape to iterate of the PE file pairs using this:

```suggestion
def _check_if_empty(fp):
    # True: empty, False: not empty, None: file doesn't exist
    if fp is None:
        return None

    with gzip.open(fp, 'rb') as fh:
        peek = fh.peek(1)
        return len(peek) == 0


def _clean_empty(seqs):
    """Clean up empty files output by Cutadapt"""
    for _, fwd_fp, rev_fp in seqs.manifest.itertuples():
        fwd_is_empty = _check_if_empty(fwd_fp)
        rev_is_empty = _check_if_empty(rev_fp)

        # case 0: both files exist and are empty
        if fwd_is_empty and rev_is_empty:
            os.remove(fwd_fp)
            os.remove(rev_fp)
        # case 1: only the fwd file exists, and is empty
        elif fwd_is_empty and rev_is_empty is None:
            os.remove(fwd_fp)
        # case 2: only the rev file exists, and is empty
        elif fwd_is_empty is None and rev_is_empty:
            os.remove(rev_fp)
        # all other cases represent situations where the files
        # shouldn't be removed
```